compilation fails with macosx 10.9 / clang

I think this is because OpenMP is not yet mature with clang on macosx. Are you using a particular compiler (new clang, gcc) on macosx?

```
> R.version
               _                           
platform       x86_64-apple-darwin13.4.0   
arch           x86_64                      
os             darwin13.4.0                
system         x86_64, darwin13.4.0        
status                                     
major          3                           
minor          1.3                         
year           2015                        
month          03                          
day            09                          
svn rev        67962                       
language       R                           
version.string R version 3.1.3 (2015-03-09)
nickname       Smooth Sidewalk   
```

```
$ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
Target: x86_64-apple-darwin13.4.0
Thread model: posix
```
```
> install_github("zarquon42b/mesheR")
Downloading github repo zarquon42b/mesheR@master
Installing mesheR
'/Library/Frameworks/R.framework/Resources/bin/R' --vanilla CMD INSTALL  \
  '/private/var/folders/2p/3fcpzn4j0313fcx185br7txw0000gn/T/RtmpuP6C7v/devtools11c25e8e435f/zarquon42b-mesheR-c46d7f7'  \
  --library='/Library/Frameworks/R.framework/Versions/3.1/Resources/library' --install-tests 

* installing *source* package ‘mesheR’ ...
** libs
clang++ -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/RcppArmadillo/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c angcheck.cpp -o angcheck.o
clang++ -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/RcppArmadillo/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c areafun.cpp -o areafun.o
clang++ -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/RcppArmadillo/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c gaussDisplace.cpp -o gaussDisplace.o
gaussDisplace.cpp:2:10: fatal error: 'omp.h' file not found
#include <omp.h>
         ^
1 error generated.
make: *** [gaussDisplace.o] Error 1
ERROR: compilation failed for package ‘mesheR’
* removing ‘/Library/Frameworks/R.framework/Versions/3.1/Resources/library/mesheR’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.1/Resources/library/mesheR’
Error: Command failed (1)
```

Hi Gregory,

actually mesheR uses the system openmp settings. To avoid this, simply set

SHLIB_OPENMP_CXXFLAGS=
in your ~/.R/Makevars file

The openmp code was a very easy way to speed up the calculation of the
smoothed displacement field and to calculate the inverse of huge sparse
block diagonal matrices.

> I think this is because OpenMP is not yet mature with clang on macosx.
> Are you using a particular compiler (new clang, gcc) on macosx?
> 
> Haha, I do not have a Mac, I use Linux.

Best
Stefan

On 24/04/15 02:02, Gregory Jefferis wrote:

> I think this is because OpenMP is not yet mature with clang on macosx.
> Are you using a particular compiler (new clang, gcc) on macosx?
> 
> |> R.version
>                _  
> platform       x86_64-apple-darwin13.4.0  
> arch           x86_64  
> os             darwin13.4.0  
> system         x86_64, darwin13.4.0  
> status  
> major          3  
> minor          1.3  
> year           2015  
> month          03  
> day            09  
> svn rev        67962  
> language       R  
> version.string R version 3.1.3 (2015-03-09)
> nickname       Smooth Sidewalk  
> |
> |$ gcc --version
> Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
> Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
> Target: x86_64-apple-darwin13.4.0
> Thread model: posix
> |
> |> install_github("zarquon42b/mesheR")
> Downloading github repo zarquon42b/mesheR@master
> Installing mesheR
> '/Library/Frameworks/R.framework/Resources/bin/R' --vanilla CMD INSTALL  \
>   '/private/var/folders/2p/3fcpzn4j0313fcx185br7txw0000gn/T/RtmpuP6C7v/devtools11c25e8e435f/zarquon42b-mesheR-c46d7f7'  \
>   --library='/Library/Frameworks/R.framework/Versions/3.1/Resources/library' --install-tests 
> - installing _source_ package ‘mesheR’ ...
>   *\* libs
>   clang++ -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/RcppArmadillo/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c angcheck.cpp -o angcheck.o
>   clang++ -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/RcppArmadillo/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c areafun.cpp -o areafun.o
>   clang++ -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/RcppArmadillo/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c gaussDisplace.cpp -o gaussDisplace.o
>   gaussDisplace.cpp:2:10: fatal error: 'omp.h' file not found
>   #include <omp.h>
>        ^
>   1 error generated.
>   make: **\* [gaussDisplace.o] Error 1
>   ERROR: compilation failed for package ‘mesheR’
> - removing ‘/Library/Frameworks/R.framework/Versions/3.1/Resources/library/mesheR’
> - restoring previous ‘/Library/Frameworks/R.framework/Versions/3.1/Resources/library/mesheR’
>   Error: Command failed (1)
>   |
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zarquon42b/mesheR/issues/1.

--- a/src/gaussDisplace.cpp
+++ b/src/gaussDisplace.cpp
@@ -1,5 +1,10 @@
 #include "RcppArmadillo.h"
+
+#include <Rconfig.h>
+#ifdef SUPPORT_OPENMP
 #include <omp.h>
+#endif
+
 using namespace Rcpp;
 using namespace arma;
 

